contracts/gateway/AxelarAmplifierGateway.sol

```diff
@@ -63,93 +78,117 @@ contract AxelarAmplifierGateway is IAxelarAmplifierGateway {
     }
 
     /**********************\
     |* External Functions *|
     \**********************/
 
-    function execute(bytes calldata batch) external {
-        (bytes memory data, bytes memory proof) = abi.decode(batch, (bytes, bytes));
+    function execute(SignedCommandBatch calldata signedBatch) external {
+        CommandBatch calldata batch = signedBatch.batch;
+        bytes memory data = abi.encode(batch);
 
-        bytes32 messageHash = ECDSA.toEthSignedMessageHash(keccak256(data));
+        // This is the EVM convention for signing non-tx data to domain separate from regular EVM txs.
+        // This should be customized per chain.
+        bytes32 batchHash = ECDSA.toEthSignedMessageHash(keccak256(data));
 
         // returns true for current operators
-        bool allowOperatorshipTransfer = authModule.validateProof(messageHash, proof);
-
-        uint256 chainId;
-        bytes32[] memory commandIds;
-        string[] memory commands;
-        bytes[] memory params;
-
-        (chainId, commandIds, commands, params) = abi.decode(data, (uint256, bytes32[], string[], bytes[]));
+        bool allowOperatorshipTransfer = authModule.validateProof(batchHash, signedBatch.proof);
 
-        if (chainId != block.chainid) revert InvalidChainId();
+        if (chainNameHash != signedBatch.batch.chainNameHash) revert InvalidChainName();
+        if (routerHash != signedBatch.batch.routerHash) revert InvalidRouter();
 
-        uint256 commandsLength = commandIds.length;
+        Command[] memory commands = batch.commands;
 
-        if (commandsLength != commands.length || commandsLength != params.length) revert InvalidCommands();
-
-        for (uint256 i; i < commandsLength; ++i) {
-            bytes32 commandId = commandIds[i];
+        for (uint256 i; i < commands.length; ++i) {
+            Command memory command = commands[i];
+            bytes32 commandId = keccak256(bytes(command.messageId));
 
             // Ignore if commandId is already executed
             if (isCommandExecuted(commandId)) {
                 continue;
             }
 
-            bytes32 commandHash = keccak256(abi.encodePacked(commands[i]));
-
-            if (commandHash == SELECTOR_APPROVE_CONTRACT_CALL) {
-                _approveContractCall(params[i], commandId);
-            } else if (commandHash == SELECTOR_TRANSFER_OPERATORSHIP) {
+            if (command.commandType == CommandType.ApproveContractCall) {
+                _approveContractCall(commandId, command);
+            } else if (command.commandType == CommandType.TransferOperatorship) {
                 if (!allowOperatorshipTransfer) {
                     continue;
                 }
 
                 allowOperatorshipTransfer = false;
 
-                _transferOperatorship(params[i]);
+                _transferOperatorship(command.params);
             } else {
-                revert InvalidCommand(commandHash);
+                revert InvalidCommand();
```

Comment:
Before we were just skipping the unknown so the batch can be resubmitted later to process the commands. (Maybe gateway upgrade, etc). Are you sure you want to revert?

Reply:
Amplifier allows re-signing approvals indefinitely unlike core (and relies commandId check to prevent replay), so commands shouldn't be stuck. We also shouldn't get unknown commands. Gateway upgrade is just a GMP approval. Original gateway had more command types with inability to re-sign.